File: scripts/prediction-sim/lib/recent-words-extractor.mjs
// Pure-logic JS port of keyboard/Sources/Prediction/Autocorrect/RecentWordsExtractor.swift.
// Kept in sync per AGENTS.md -> Test policy.

import { canonicalize } from './text-normalization.mjs';

// Unicode property check for "is punctuation".
// Matches Swift's Character.isPunctuation for the characters we care about
// (quotes, apostrophes, brackets, commas, periods, etc.).
const PUNCTUATION_RE = /[\p{P}]/u;

function isPunctuation(ch) {
  return PUNCTUATION_RE.test(ch);
}

/**
 * Strip ALL trailing punctuation (apostrophes included) from a finished word.
 * Returns null if the input is null/empty or becomes empty after stripping.
 *
 * Local copy mirroring the private helper in RecentWordsExtractor.swift, which
 * itself mirrors CurrentWordExtractor.swift — the Swift helpers are private per
 * file, so the ports carry the same per-file copy.
 */
function stripTrailingPunctuation(word) {
  if (word === null || word === undefined || word === '') return null;
  let result = word;
  while (result.length > 0 && isPunctuation(result[result.length - 1])) {
    result = result.slice(0, -1);
  }
  return result === '' ? null : result;
}

function isWhitespace(ch) {
  return /\s/.test(ch);
}

/**
 * Extracts recently committed words walking backwards from the cursor.
 *
 * Mirrors RecentWordsExtractor.extract(from:maxCount:) in Swift.
 *
 * @param {string|null|undefined} context - The full string before the cursor.
 * @param {number} maxCount - Maximum number of committed words to return (default 4).
 * @returns {Array<{word: string, lookupWord: string, offsetFromCursorEnd: number}>}
 *   Most recent word first. `offsetFromCursorEnd` is the char distance from the
 *   cursor to the END of the word's body (after trailing punctuation).
 */
export function extract(context, maxCount = 4) {
  if (context === null || context === undefined || context === '') return [];
  if (!(maxCount > 0)) return [];

  const endsWithWhitespace = /\s$/.test(context);

  const results = [];
  let index = context.length;

  // Skip trailing whitespace so `index` lands just after the last token.
  while (index > 0 && isWhitespace(context[index - 1])) {
    index--;
  }

  // If the cursor is mid-word, skip the in-progress token and the whitespace
  // before it, so `index` points just after the last committed word.
  if (!endsWithWhitespace) {
    while (index > 0) {
      const previous = context[index - 1];
      if (isWhitespace(previous)) {
        index--;
        break;
      }
      index--;
    }
  }

  // Collect committed words walking backwards, up to maxCount.
  while (results.length < maxCount && index > 0) {
    // Find the start of the token ending at `index`.
    let tokenStart = index;
    while (tokenStart > 0 && !isWhitespace(context[tokenStart - 1])) {
      tokenStart--;
    }

    const word = context.slice(tokenStart, index);
    const stripped = stripTrailingPunctuation(word) ?? word;
    const lookupWord = canonicalize(stripped);

    // The word's body (after trailing punctuation) ends `strippedCount`
    // characters before the token's display end. The apply plan deletes exactly
    // the body, so the offset must point at the body end.
    const strippedCount = word.length - stripped.length;
    const bodyEnd = index - strippedCount;

    results.push({
      word,
      lookupWord,
      offsetFromCursorEnd: context.length - bodyEnd,
    });

    // Advance to the previous token: hop over the current token's start
    // position, then over any whitespace separating it from the next token.
    index = tokenStart;
    while (index > 0 && isWhitespace(context[index - 1])) {
      index--;
    }
  }

  return results;
}

File: keyboard/Sources/Prediction/Autocorrect/RecentWordsExtractor.swift
import Foundation

/// Pure function for extracting recently committed words from the document
/// context, walking backwards from the cursor. No UIKit import.
///
/// Used by retroactive autocorrect: when the user presses a separator, the
/// keyboard can revisit the last few words and correct ones that were typed
/// without an inline suggestion.
enum RecentWordsExtractor {

    /// A single committed word found before the cursor.
    struct RecentWord {
        /// The token as it appears in the document (trailing punctuation
        /// preserved, for display).
        let word: String
        /// The token with trailing punctuation stripped and apostrophes
        /// canonicalized (for dictionary lookups).
        let lookupWord: String
        /// Character distance from the cursor to the END of the word's body
        /// (after trailing punctuation). `RetroactiveApplyPlan` navigates back
        /// to the word with `backMove == -offsetFromCursorEnd`.
        let offsetFromCursorEnd: Int
    }

    /// Walks backwards from the cursor (the end of `context`), collecting up to
    /// `maxCount` completed words.
    ///
    /// If the cursor is mid-word, the in-progress token is skipped — only
    /// committed words can be retroactively corrected. The most recent word is
    /// returned first; each entry's `offsetFromCursorEnd` grows as the walk
    /// moves further back.
    ///
    /// Tokenization mirrors `CurrentWordExtractor`: whitespace-delimited
    /// tokens, trailing punctuation stripped from finished words, apostrophes
    /// canonicalized via `ApostropheNormalizer.canonicalize`.
    ///
    /// - Parameters:
    ///   - context: The full string before the cursor (from `textDocumentProxy.documentContextBeforeInput`).
    ///   - maxCount: Maximum number of committed words to return (older words are dropped).
    /// - Returns: Up to `maxCount` `RecentWord`s, most recent first.
    static func extract(from context: String?, maxCount: Int) -> [RecentWord] {
        guard let context, !context.isEmpty, maxCount > 0 else { return [] }

        // True when the cursor sits at a word boundary (the last char is whitespace).
        let endsWithWhitespace = context.last?.isWhitespace ?? true

        var results: [RecentWord] = []
        results.reserveCapacity(maxCount)

        var index = context.endIndex

        // Skip trailing whitespace so `index` lands just after the last token.
        while index > context.startIndex {
            let previous = context.index(before: index)
            if context[previous].isWhitespace {
                index = previous
            } else {
                break
            }
        }

        // If the cursor is mid-word, the token ending at `index` is the word
        // being typed — skip it and the whitespace before it, so `index` points
        // just after the last committed word. Only committed words qualify.
        if !endsWithWhitespace {
            while index > context.startIndex {
                let previous = context.index(before: index)
                if context[previous].isWhitespace {
                    index = previous
                    break
                }
                index = previous
            }
        }

        // Collect committed words walking backwards, up to maxCount.
        while results.count < maxCount, index > context.startIndex {
            // Find the start of the token ending at `index`.
            var tokenStart = index
            while tokenStart > context.startIndex {
                let previous = context.index(before: tokenStart)
                if context[previous].isWhitespace {
                    break
                }
                tokenStart = previous
            }

            let word = String(context[tokenStart..<index])
            let stripped = stripTrailingPunctuation(from: word) ?? word
            let lookupWord = ApostropheNormalizer.canonicalize(stripped)

            // The word's body (after trailing punctuation) ends `strippedCount`
            // characters before the token's display end. The apply plan deletes
            // exactly the body, so the offset must point at the body end.
            let strippedCount = word.count - stripped.count
            let bodyEnd = context.index(index, offsetBy: -strippedCount)

            results.append(RecentWord(
                word: word,
                lookupWord: lookupWord,
                offsetFromCursorEnd: context[bodyEnd..<context.endIndex].utf16.count
            ))

            // Advance to the previous token: hop over the current token's start
            // position, then over any whitespace separating it from the next token.
            index = tokenStart
            while index > context.startIndex {
                let previous = context.index(before: index)
                if context[previous].isWhitespace {
                    index = previous
                } else {
                    break
                }
            }
        }

        return results
    }

    // MARK: - Helpers

    /// Strip ALL trailing punctuation (apostrophes included) from a finished
    /// word. Returns `nil` if the input is empty or becomes empty after
    /// stripping. Mirrors `CurrentWordExtractor.stripTrailingPunctuation(from:)`
    /// (private there, so duplicated here).
    private static func stripTrailingPunctuation(from word: String) -> String? {
        guard !word.isEmpty else { return nil }
        var result = word
        while let last = result.last, last.isPunctuation {
            result = String(result.dropLast())
        }
        return result.isEmpty ? nil : result
    }
}
